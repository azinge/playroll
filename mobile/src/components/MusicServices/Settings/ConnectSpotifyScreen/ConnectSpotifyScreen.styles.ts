import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  container: {
    flex: 1,
    height: '100%',
    width: '100%',
    alignItems: 'center',
  },
  spotifyIcon: {
    marginTop: 150,
    marginBottom: 20,
    height: 200,
    width: 200,
  },
  connectButton: {
    marginVertical: 10,
    borderRadius: 10,
    backgroundColor: '#A753AA',
  },
});
