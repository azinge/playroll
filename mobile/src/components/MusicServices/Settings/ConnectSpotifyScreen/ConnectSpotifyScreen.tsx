/**
 * ConnectSpotifyScreen
 */

import * as React from 'react';
import { Text, SafeAreaView, Image } from 'react-native';
import { Button } from 'react-native-elements';
import { WebBrowser } from 'expo';
import url from 'url';

import styles from './ConnectSpotifyScreen.styles';
import { RegisterSpotifyAuthCodeMutation } from '../../../../graphql/requests/Spotify/RegisterSpotifyAuthCodeMutation';
import SubScreenContainer from '../../../shared/Containers/SubScreenContainer';

export default class ConnectSpotifyScreen extends React.Component {
  async handleOpenSpotifyAuthView(): Promise<string> {
    const scope =
      'user-read-private+user-library-read+user-read-email+playlist-modify-public+playlist-modify-private';
    const responseType = 'code';
    const clientID = 'e5149b4616b84918911f9419a279d23b';
    const redirectURI = 'http://app.playroll.io';
    const authParams = `client_id=${clientID}&redirect_uri=${redirectURI}&response_type=${responseType}&scope=${scope}`;
    const uri = `https://accounts.spotify.com/authorize?${authParams}`;
    let result = await WebBrowser.openAuthSessionAsync(uri);
    if (result.type === 'success') {
      const obj = url.parse(result.url, true);
      return obj.query.code;
    }
    throw 'Error connecting to spotify';
  }

  render() {
    return (
      <SubScreenContainer title='Connect to Spotify'>
        <Text>ConnectSpotifyScreen</Text>
        <Image
          style={styles.spotifyIcon}
          source={require('../../../../assets/spotifyIcon.png')}
        />
        <RegisterSpotifyAuthCodeMutation>
          {registerSpotifyAuthCode => {
            return (
              <Button
                title='Connect to Spotify'
                containerStyle={styles.connectButton}
                onPress={async () => {
                  try {
                    const code = await this.handleOpenSpotifyAuthView();
                    registerSpotifyAuthCode({ variables: { code } });
                  } catch (err) {
                    console.log(err);
                  }
                }}
              />
            );
          }}
        </RegisterSpotifyAuthCodeMutation>
      </SubScreenContainer>
    );
  }
}
