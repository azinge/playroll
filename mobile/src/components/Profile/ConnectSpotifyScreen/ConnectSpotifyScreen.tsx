/**
 * ConnectSpotifyScreen
 */

import * as React from "react";
import { Text, View, SafeAreaView } from "react-native";

export default class ConnectSpotifyScreen extends React.Component {
  render() {
    return (
      <SafeAreaView>
        <Text>ConnectSpotifyScreen</Text>
      </SafeAreaView>
    );
  }
}
