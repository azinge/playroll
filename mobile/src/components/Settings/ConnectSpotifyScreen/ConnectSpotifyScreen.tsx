/**
 * ConnectSpotifyScreen
 */

import * as React from 'react';
import { Text, SafeAreaView, Image, View } from 'react-native';
import { Button } from 'react-native-elements';
import { Linking, WebBrowser } from 'expo';
import url from 'url';

import styles from './ConnectSpotifyScreen.styles';
import { RegisterSpotifyAuthCodeMutation } from '../../../graphql/requests/Spotify/RegisterSpotifyAuthCodeMutation';
import SubScreenContainer from '../../shared/Containers/SubScreenContainer';
import DropdownAlert from 'react-native-dropdownalert';
import { CurrentUserSpotifyStatusQuery } from '../../../graphql/requests/Spotify';
import NavigationService from '../../../services/NavigationService';

export type Props = {};

type State = {
  code: string;
};

export default class ConnectSpotifyScreen extends React.Component<
  Props,
  State
> {
  dropdown: DropdownAlert;

  constructor(props: Props) {
    super(props);

    this.state = {
      code: '',
    };

    this.handleRedirect = this.handleRedirect.bind(this);
  }

  handleOpenSpotifyAuthView() {
    const scope =
      'user-read-private+user-library-read+user-read-email+playlist-modify-public+playlist-modify-private';
    const responseType = 'code';
    const clientID = 'e5149b4616b84918911f9419a279d23b';
    // const redirectURI = `http://app.playroll.io`;
    const redirectURI = `https://app-dev.playroll.io`;
    const authParams = `client_id=${clientID}&redirect_uri=${redirectURI}&response_type=${responseType}&scope=${scope}`;
    const uri = `https://accounts.spotify.com/authorize?${authParams}`;
    WebBrowser.openBrowserAsync(uri);
  }

  handleRedirect(result) {
    console.log(result);
    if (result.url) {
      WebBrowser.dismissBrowser();
      const obj = url.parse(result.url, true);
      this.setState({ code: obj.query.code });
    } else {
      throw 'Error connecting to spotify';
    }
  }

  componentDidMount() {
    Linking.addEventListener('url', this.handleRedirect);
  }
  componentWillUnmount() {
    Linking.removeEventListener('url', this.handleRedirect);
  }

  render() {
    return (
      <View style={{ flex: 1 }}>
        <SubScreenContainer title='Connect to Spotify' modal>
          <View style={{ alignItems: 'center' }}>
            <Image
              style={styles.spotifyIcon}
              source={require('../../../assets/spotifyIcon.png')}
            />
            <RegisterSpotifyAuthCodeMutation
              onCompleted={() => {
                setTimeout(() => NavigationService.goBack(), 1000);
              }}
              onError={err => {
                console.log(err);
              }}
            >
              {(registerSpotifyAuthCode, { loading }) => {
                const { code: stateCode } = this.state;
                if (stateCode !== '') {
                  this.setState({ code: '' }, async () => {
                    await registerSpotifyAuthCode({
                      variables: { code: stateCode, devMode: true },
                      // variables: { code: stateCode },
                    });
                  });
                }
                return (
                  <Button
                    title='Connect to Spotify'
                    titleStyle={{ fontWeight: 'bold' }}
                    containerStyle={[
                      styles.connectButton,
                      {
                        borderRadius: 80,
                        width: '75%',
                        height: 50,
                      },
                    ]}
                    onPress={() => {
                      this.handleOpenSpotifyAuthView();
                    }}
                    buttonStyle={{
                      borderRadius: 80,
                      height: 50,
                      backgroundColor: '#af00bc',
                    }}
                    raised
                    loading={loading}
                  />
                );
              }}
            </RegisterSpotifyAuthCodeMutation>
          </View>
        </SubScreenContainer>
        <DropdownAlert ref={ref => (this.dropdown = ref)} />
      </View>
    );
  }
}
