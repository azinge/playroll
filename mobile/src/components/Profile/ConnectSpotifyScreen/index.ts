import ConnectSpotifyScreen from "./ConnectSpotifyScreen";
export default ConnectSpotifyScreen;
