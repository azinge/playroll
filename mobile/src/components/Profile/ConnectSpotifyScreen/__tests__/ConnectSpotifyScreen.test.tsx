import * as React from 'react';
import renderer from 'react-test-renderer';

import ConnectSpotifyScreen from '../ConnectSpotifyScreen';

it('renders correctly', () => {
  const tree = renderer.create(<ConnectSpotifyScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});
