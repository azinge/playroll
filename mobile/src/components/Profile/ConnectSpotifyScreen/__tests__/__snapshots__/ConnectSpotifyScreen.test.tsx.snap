// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <Text>
    ConnectSpotifyScreen
  </Text>
</RCTSafeAreaView>
`;
